source/Calamari.Tests/Fixtures/Integration/FileSystem/GlobExpression/CharGroupRetrieverFixture.cs

```diff
@@ -0,0 +1,54 @@
+using System;
+using System.Linq;
+using Calamari.Common.Plumbing.FileSystem.GlobExpressions;
+using FluentAssertions;
+using Newtonsoft.Json;
+using NUnit.Framework;
+using Octopus.CoreUtilities.Extensions;
+
+namespace Calamari.Tests.Fixtures.Integration.FileSystem.GlobExpression
+{
+    [TestFixture]
+    public class CharGroupRetrieverFixture
+    {
+        [Test]
+        // Returns no groups
+        [TestCase("hello.txt", "")] // No {} groups
+        [TestCase("Here/[hello.txt", "")] // dangling [
+        [TestCase("Her[e/h]ello.txt", "")] // [] separated by directory separator
+        [TestCase("Here/[te{s]t,mp}.txt", "")] // [] intersects with {}
+
+        // Returns groups
+        [TestCase("Here/[hi].txt", "{startIndex:5,length:4,options:['h','i','[hi]']}")] // Basic char group, will also return literal expression.
+        [TestCase("Here/[a-c].txt", "{startIndex:5,length:5,options:['a','b','c']}")] // Range char Group
+        [TestCase("This/I[st]/Fi[rn]e.png",
+            "{startIndex:6,length:4,options:['s','t','[st]']};" +
+            "{startIndex:13,length:4,options:['r','n','[rn]']}")] // Two char groups
+        [TestCase("Here/[?*.].txt", "{startIndex:5,length:5,options:['?','*','.','[?*.]']}")] // Other characters
```

Comment:
It's a minor nit but by putting a JSON string here, you lose the connection to the code model definition.. Did you considering something like `[TestCaseSource]` and having that return pre-defined `Group` objects.

The tests are really good, i'm just concerned having the JSON strings are your test cases

Reply:
My aim with using JSON like this was mainly brevity. I also find it a bit more readable compared to using a TestCaseSource implementation.

I've updated the code to use a TestCaseSource but to my eye it's harder to parse quickly because of the required boilerplate in C# compared to writing the class in json.

Example:
![image](https://github.com/OctopusDeploy/Calamari/assets/97430840/64397613-f0f8-4ce7-b790-4cb7b048bbe3)

I agree it disconnects a bit from the `Group` model and won't show compile time errors if someone updates the model which is a pain. I'm happy to concede here but I wish there were a better, cleaner option!

Reply:
Yeah, I can understand where you are coming from. I think it's ok to lose a bit of brevity to gain the compiler support.
There could be some extra stuff we could do to make the test cases themselves easier to read if you want to play with it? I'm not fussed at this point

Reply:
In the end I agree and have come around because of the case where I would prefer the compiler complain than to find out tests have broken after the fact.